### PR TITLE
8320362: Load anchor certificates from Keychain keystore

### PR DESCRIPTION
**Pull request by alexeybakhtin**
Please review the proposed fix.

The patch loads system root certificates from the MacOS Keychain with TrustSettings.
It allows to build a trusted certificate path using the MacOS Keychain store only.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8321045](https://bugs.openjdk.org/browse/JDK-8321045) to be approved

### Issues
 * [JDK-8320362](https://bugs.openjdk.org/browse/JDK-8320362): Load anchor certificates from Keychain keystore (**Enhancement** - P4)
 * [JDK-8321045](https://bugs.openjdk.org/browse/JDK-8321045): Load anchor certificates from Keychain keystore (**CSR**)

### Reviewers
 * @1keep2keepFaith (no known openjdk.org user name / role) ⚠️ Review applies to [db98b7c2](https://git.openjdk.org/jdk/pull/16722/files/db98b7c2ed0250edb13bf54167d207f57ab7ac2a)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to [24ca1206](https://git.openjdk.org/jdk/pull/16722/files/24ca1206b3ba1d5dfdbab2e1042108b6543a4283)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to [b92e21b9](https://git.openjdk.org/jdk/pull/16722/files/b92e21b97c680e77e76b86336200631a641cdbfb)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16722/head:pull/16722` \
`$ git checkout pull/16722`

Update a local copy of the PR: \
`$ git checkout pull/16722` \
`$ git pull https://git.openjdk.org/jdk.git pull/16722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16722`

View PR using the GUI difftool: \
`$ git pr show -t 16722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16722.diff">https://git.openjdk.org/jdk/pull/16722.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16722#issuecomment-1818342769)

### CODE REVIEW

**Comment by 1keep2keepFaith** [comment]
Thanks please let me know if I need to do other

**Comment by alexeybakhtin** [comment]
As suggested by @wangweij, the new Keychain-ROOT keystore is introduced for the trusted anchor certificates.
The Keychain-ROOT keystore is read-only and throws KeyStoreException in an attempt to modification

**Comment by alexeybakhtin** [comment]
> What are the change for existing `addCertificatesToKeystore` for? Is there any behavior change?

Hi @wangweij . No behavior changes. Just reformatted to make it similar to addCertificatesToKeystoreRoot. Can be reverted back.

File: src/java.base/macosx/classes/apple/security/AppleProvider.java [hunk 1]
```diff
@@ -82,7 +84,9 @@ public AppleProvider() {
         AccessController.doPrivileged(new PrivilegedAction<Void>() {
             public Void run() {
                 putService(new ProviderService(p, "KeyStore",
-                           "KeychainStore", "apple.security.KeychainStore"));
+                           "KeychainStore", "apple.security.KeychainStore.USER"));
+                putService(new ProviderService(p, "KeyStore",
+                           "KeychainStore-ROOT", "apple.security.KeychainStore.ROOT"));
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/classes/apple/security/AppleProvider.java]: Thank you. updated with the $ sign

File: src/java.base/macosx/classes/apple/security/KeychainStore.java [hunk 2]
```diff
@@ -42,12 +42,72 @@
...
- * a file-based implementation.
+ * This class provides the keystores implementation referred to as
+ * "KeychainStore" and "KeychainStore-ROOT".
+ * It uses the current user's and system root keychains accordingly
+ * as its backing storage, and does NOT support a file-based
+ * implementation.
  */
 
-public final class KeychainStore extends KeyStoreSpi {
+abstract class KeychainStore extends KeyStoreSpi {
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/classes/apple/security/KeychainStore.java]: Good catch. Thank you. Added `sealed`

File: src/java.base/macosx/classes/apple/security/KeychainStore.java [hunk 3]
```diff
@@ -137,14 +197,25 @@ private static void permissionCheck() {
...
-    public KeychainStore() { }
+    private KeychainStore(String name) {
+        this.storeName = name;
+    }
+
+    /**
+     * Returns the name of the keystore.
+     */
+    private String getName()
+    {
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/classes/apple/security/KeychainStore.java]: OK. Not much reason for this method. removed

File: src/java.base/macosx/classes/apple/security/KeychainStore.java [hunk 4]
```diff
@@ -472,6 +543,7 @@ public void engineSetKeyEntry(String alias, byte[] key,
         permissionCheck();
 
         synchronized(entries) {
+
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/classes/apple/security/KeychainStore.java]: Fixed

File: src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m [hunk 5]
```diff
@@ -471,32 +471,21 @@ static void addCertificatesToKeystore(JNIEnv *env, jobject keyStore)
...
-
-            // Load user trustSettings into inputTrust
-            if (SecTrustSettingsCopyTrustSettings(certRef, kSecTrustSettingsDomainUser, &trustSettings) == errSecSuccess && trustSettings != NULL) {
-                inputTrust = (*env)->NewObject(env, jc_arrayListClass, jm_arrayListCons);
-                if (inputTrust == NULL) {
+            for (SecTrustSettingsDomain domain = kSecTrustSettingsDomainUser; domain <= kSecTrustSettingsDomainAdmin; domain++) {
+                trustSettings = NULL;
+                // Load user trustSettings into inputTrust
+                if (SecTrustSettingsCopyTrustSettings(certRef, domain, &trustSettings) == errSecSuccess && trustSettings != NULL) {
+                    if(inputTrust == NULL) {
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m]: Fixed

File: src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m [hunk 6]
```diff
@@ -519,6 +508,102 @@ static void addCertificatesToKeystore(JNIEnv *env, jobject keyStore)
...
+        goto errOut;
+    }
+
+    jmethodID jm_listAdd = (*env)->GetMethodID(env, jc_arrayListClass, "add", "(Ljava/lang/Object;)Z");
+    if (jm_listAdd == NULL) {
+        goto errOut;
+    }
+
+    // Read Trust Anchors
+    if(SecTrustCopyAnchorCertificates(&currAnchors) == errSecSuccess) {
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m]: Fixed

**Comment by alexeybakhtin** [comment]
> Is it possible to reuse some some lines from `addCertificatesToKeystore`?
> 
> BTW, I reviewed the CSR.

Hi @wangweij,
Thank you a lot for PR and CSR review. I have updated PR with review findings and refactored addCertificatesToKeystore/addCertificatesToKeystoreRoot to reuse some code

**Comment by alexeybakhtin** [comment]
> Do you intend to call `addIdentitiesToKeystore` for both store types?

Yes, you are right. Thank you. We do not need private identities in the root keystore. Updated

**Comment by alexeybakhtin** [comment]
> Great! The change looks good. Can you manage to add a test? Maybe try to load both store types. Make sure they have different contents and not empty (?).

Thank you. I have updated `java/security/KeyStore/CheckMacOSKeyChainTrust.java` test to load system root certificates and compare it with the content of the `KEYCHAINSTORE-ROOT` keystore.

File: test/jdk/java/security/KeyStore/CheckMacOSKeyChainTrust.java [hunk 7]
```diff
@@ -32,22 +32,32 @@
...
-        loadUser();
-        loadAdmin();
+        String keystore = args[0];
+        if (keystore.equals("KEYCHAINSTORE")) {
+            loadUser(true);
+            loadAdmin(true);
+        } else {
+            // check user and admin trustsettings to find distrusted certs
+            loadUser(false);
+            loadAdmin(false);
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply test/jdk/java/security/KeyStore/CheckMacOSKeyChainTrust.java:55]: Yes. Exactly. The trusted cert can be distrusted in the user/admin domain, so It should not be available in the KEYCHAINSTORE-ROOT

**Comment by alexeybakhtin** [comment]
> Will KEYCHAINSTORE-ROOT contains trusted certs inside KEYCHAINSTORE? I tried on my machine and there are some items not in `security dump-trust-settings -s`.

I've found an issue in the original implementation that prevents adding different certificates with the same CN. I am going to fix it as part of this PR because it affects the scope of root certificates in the KEYCHAINSTORE-ROOT

**Comment by alexeybakhtin** [comment]
> Will KEYCHAINSTORE-ROOT contains trusted certs inside KEYCHAINSTORE? I tried on my machine and there are some items not in `security dump-trust-settings -s`.
`security dump-trust-settings -s` returns only predefined root certificates. KEYCHAINSTORE-ROOT additionally contains installed root trusted certificates in the system domain

File: test/jdk/java/security/KeyStore/CheckMacOSKeyChainTrust.java [hunk 8]
```diff
@@ -32,22 +32,32 @@
...
 /*
  * @test
- * @bug 8303465
+ * @bug 8303465 8320362
  * @library /test/lib
  * @requires os.family == "mac"
  * @summary Check whether loading of certificates from MacOS Keychain correctly
  *          honors trust settings
+ * @run main CheckMacOSKeyChainTrust KEYCHAINSTORE
+ * @run main CheckMacOSKeyChainTrust KEYCHAINSTORE-ROOT
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply test/jdk/java/security/KeyStore/CheckMacOSKeyChainTrust.java:41]: Yes, we can, but it looks like not required until we compare the contents of these stores

**Comment by alexeybakhtin** [comment]
> > `security dump-trust-settings -s` returns only predefined root certificates. KEYCHAINSTORE-ROOT additionally contains installed root trusted certificates in the system domain
> 
> Are you sure they should be added into this keystore? It looks like all the extra certs in KEYCHAINSTORE-ROOT that are not in `security dump-trust-settings -s` are all inside KEYCHAINSTORE. Maybe that's where they should belong to?

Thank you. You are right. It is better if KEYCHAINSTORE-ROOT contains only predefined roots. Unfortunately, SecTrustCopyAnchorCertificates can not be used in this case.

I have updated the patch to read certificates from the "/System/Library/Keychains/SystemRootCertificates.keychain" keychain

**Comment by alexeybakhtin** [comment]
@wangweij could you please take a look at the last proposed commit, if you have any chance

File: src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m [hunk 9]
```diff
@@ -515,10 +515,32 @@ static void addCertificatesToKeystoreRoot(JNIEnv *env, jobject keyStore,
...
+    SecKeychainRef keychain = NULL;
+    CFMutableArrayRef keychainList = NULL;
+    CFDictionaryRef search = NULL;
     CFArrayRef currAnchors = NULL;
 
-    // Read Trust Anchors
-    if (SecTrustCopyAnchorCertificates(&currAnchors) == errSecSuccess) {
+    // Load predefined root certificates from SystemRootCertificates keychain
+    // SecTrustCopyAnchorCertificates includes extra root certificates and can not be used here
+    if( SecKeychainOpen("/System/Library/Keychains/SystemRootCertificates.keychain", &keychain) != errSecSuccess ) {
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:525]: Hi @wangweij , 
Thank you for review.
Unfortunately `SecTrustCopyCustomAnchorCertificates` can not be used also. It is used to retrieve certificates from your own created SecTrust. As I know it is not possible to create/load SecTrust with predefined certs without reading`/System/Library/Keychains/SystemRootCertificates.keychain`

[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:525]: Thank you. I've updated CSR and waiting for review.

[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:525]: CSR is in the Proposed state now. Workflow does not allow me to Finalize it.

[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m:525]: Thank you. Moved to Finalized

**Comment by seanjmullan** [comment]
Is it practical to add a test as described in the bug?:

`java -Djavax.net.ssl.trustStoreType=KeychainStore-ROOT HttpsURLConnectionTest https://github.com/`

This could be added to the `test/jdk/security/infra` test directory since it depends on an external server.

**Comment by alexeybakhtin** [comment]
Hi @seanjmullan

**Comment by seanjmullan** [comment]
> Hi @seanjmullan Thank you for review I've added the test

Thanks. @rhalade Is this an acceptable place (security/infra) to put a test that makes external network connections?

**Comment by rhalade** [comment]
> > Hi @seanjmullan Thank you for review I've added the test
> 
> Thanks. @rhalade Is this an acceptable place (security/infra) to put a test that makes external network connections?

Yes, it is correct place to add this infra test. @alexeybakhtin please update https://github.com/openjdk/jdk/blob/master/test/jdk/TEST.groups#L254 definition to include this test. Update could be:

jdk_security_infra = \
    security/infra

**Review by seanjmullan — APPROVED** [review]
I only reviewed HttpsURLConnectionTest.java and had one minor comment (typo).

File: test/jdk/security/infra/javax/net/ssl/HttpsURLConnectionTest.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8320362
+ * @summary Verifies successfull connection to external server with
```
[the remark this answers is not in the document]
Comment (alexeybakhtin) [reply test/jdk/security/infra/javax/net/ssl/HttpsURLConnectionTest.java]: Thank you. Fixed

**Pull request closed without merging**